src/main/java/com/palantir/common/streams/BackpressureSpliterator.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.common.streams;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.Spliterator;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+class BackpressureSpliterator<U> implements Spliterator<U> {
+    private final int desiredParallelism;
+    private final BlockingQueue<CompletableFuture<U>> completed;
+    private final Spliterator<CompletableFuture<U>> notStarted;
+
+    private int inProgress = 0;
+
+    private BackpressureSpliterator(
+            int desiredParallelism,
+            Spliterator<CompletableFuture<U>> arguments) {
+        checkArgument(desiredParallelism > 0, "desiredParallelism %s > 0", new Object[] { desiredParallelism });
```

Comment:
ouch

Reply:
"maxParallelism must be at least 1 (got %s)"?

Reply:
done.